tools/elfdeps.c

```diff
@@ -97,6 +99,102 @@ static const char *mkmarker(GElf_Ehdr *ehdr)
     return marker;
 }
 
+static const char *mkarchmarker(GElf_Ehdr *ehdr)
+{
+    /* Hopefully this is enough for the arch marker */
+    char archmarker[25];
+    const char* elf_machine = NULL;
+    const char* elf_endian = NULL;
+    const char* elf_bitsize = NULL;
+
+    /* First the machine type... */
+    switch (ehdr->e_machine) {
+    case EM_SPARC:
+	elf_machine = "sparc";
+	break;
+    case EM_386:
+    case EM_X86_64:
+	elf_machine = "x86";
+	break;
+    case EM_MIPS:
+	elf_machine = "mips";
+	break;
+    case EM_PPC:
+    case EM_PPC64:
+	elf_machine = "ppc";
+	break;
+    case EM_S390:
+	elf_machine = "s390";
+	break;
+    case EM_ARM:
+	if ((ehdr->e_flags | EF_ARM_ABI_FLOAT_HARD) == EF_ARM_ABI_FLOAT_HARD)
+		elf_machine = "armhfp";
+	if ((ehdr->e_flags | EF_ARM_ABI_FLOAT_SOFT) == EF_ARM_ABI_FLOAT_SOFT)
+		elf_machine = "armsfp";
+	break;
+    case EM_SPARCV9:
+	elf_machine = "sparcv9";
+	break;
+    case EM_ALPHA:
+	elf_machine = "alpha";
+	break;
+    case EM_AARCH64:
+	elf_machine = "aarch";
+	break;
+    case EM_RISCV:
+	elf_machine = "riscv";
+	break;
+    default:
+	break;
+    }
+
+    /* Then the endianness of the CPU... */
+    switch (ehdr->e_ident[EI_DATA]) {
+    case ELFDATA2LSB:
+	elf_endian = "le";
+	break;
+    case ELFDATA2MSB:
+	elf_endian = "be";
+	break;
+    default:
+	break;
+    }
+
+    /* Then the bit size of the CPU... */
+    switch (ehdr->e_ident[EI_CLASS]) {
+    case ELFCLASS64:
+	elf_bitsize = "64";
+	break;
+    case ELFCLASS32:
+	elf_bitsize = "32";
+	break;
+    default:
+	break;
+    }
+
+    /* Finally the arch marker! */
+    switch (ehdr->e_machine) {
+    case EM_MIPS:
+    case EM_ARM:
+    case EM_PPC64:
+	snprintf(archmarker, sizeof(archmarker), "(%s%s-%s)", elf_machine, elf_endian, elf_bitsize);
+	break;
+    case EM_X86_64:
+	/* This handling for x32 makes me weep inside... */
+	if (ehdr->e_ident[EI_CLASS] == ELFCLASS32) {
+	    snprintf(archmarker, sizeof(archmarker), "(%s-%s-%s)", elf_machine, "64", "x32");
+	} else {
+	    snprintf(archmarker, sizeof(archmarker), "(%s-%s)", elf_machine, elf_bitsize);
+	}
+	break;
+    default:
+	snprintf(archmarker, sizeof(archmarker), "(%s-%s)", elf_machine, elf_bitsize);
+	break;
+    }
+
+    return strndup(archmarker, strlen(archmarker));
```

Comment:
Um, why the static-sized "lets hope this is enough" buffer when you end up strdup()'ing it afterall?
Just use rasprintf() to construct the marker to begin with...

Reply:
@pmatilai Because I've never heard of `rasprintf()` before? Any reference to it somewhere?

Reply:
rasprintf() is rpm's local version of asprintf() (so we can rely on it always being there), it's used all over rpm codebase -  elfdeps.c included.